hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/receiver/kafka/KafkaSingleThreadedMessageReceiver.java

```diff
@@ -0,0 +1,162 @@
+                .collect(Collectors.toMap(t -> t.name().asString(), Function.identity()));
+        this.consumer = consumer;
+        this.messageContentWrapper = messageContentWrapper;
+        this.schemaRepository = schemaRepository;
+        this.clock = clock;
+        this.readQueue = new ArrayBlockingQueue<Message>(readQueueCapacity);
+        this.consumer.subscribe(topics.keySet());
+    }
+
+    private Collection<KafkaTopic> getKafkaTopics(Topic topic, KafkaNamesMapper kafkaNamesMapper) {
+        KafkaTopics kafkaTopics = kafkaNamesMapper.toKafkaTopics(topic);
+        ImmutableList.Builder<KafkaTopic> topicsBuilder = new ImmutableList.Builder<KafkaTopic>().add(kafkaTopics.getPrimary());
+        kafkaTopics.getSecondary().ifPresent(topicsBuilder::add);
+        return topicsBuilder.build();
+    }
+
+    @Override
+    public Optional<Message> next() {
+        try {
+            if (readQueue.isEmpty()) {
+                ConsumerRecords<byte[], byte[]> records = consumer.poll(pollTimeout);
+                for (ConsumerRecord<byte[], byte[]> record : records) {
+                    readQueue.add(convertToMessage(record));
+                }
+            }
+            return Optional.ofNullable(readQueue.poll());
+        } catch (KafkaException ex) {
+            logger.error("Error while reading message for subscription {}", subscription.getQualifiedName(), ex);
+            return Optional.empty();
+        }
+    }
+
+    private Message convertToMessage(ConsumerRecord<byte[], byte[]> record) {
+        UnwrappedMessageContent unwrappedContent = getUnwrappedMessageContent(record);
+        KafkaTopic kafkaTopic = topics.get(record.topic());
+        return new Message(
+                unwrappedContent.getMessageMetadata().getId(),
+                topic.getQualifiedName(),
+                unwrappedContent.getContent(),
+                kafkaTopic.contentType(),
+                unwrappedContent.getSchema(),
+                unwrappedContent.getMessageMetadata().getTimestamp(),
+                clock.millis(),
+                new PartitionOffset(kafkaTopic.name(), record.offset(), record.partition()),
+                unwrappedContent.getMessageMetadata().getExternalMetadata(),
+                subscription.getHeaders()
+        );
+    }
+
+    private UnwrappedMessageContent getUnwrappedMessageContent(ConsumerRecord<byte[], byte[]> message) {
+        if (topic.getContentType() == ContentType.AVRO) {
+            return messageContentWrapper.unwrapAvro(message.value(), topic, schemaRepository);
+        } else if (topic.getContentType() == ContentType.JSON) {
+            return messageContentWrapper.unwrapJson(message.value());
+        }
+        throw new UnsupportedContentTypeException(topic);
+    }
+
+    @Override
+    public void stop() {
+        consumer.close();
+    }
+
+    @Override
+    public void update(Subscription newSubscription) {
+        this.subscription = subscription;
```

Comment:
\[FindBugs\] WARNING: SA: Self assignment of field KafkaSingleThreadedMessageReceiver\.subscription in pl\.allegro\.tech\.hermes\.consumers\.consumer\.receiver\.kafka\.KafkaSingleThreadedMessageReceiver\.update\(Subscription\)